src/validity.cpp

```diff
@@ -763,7 +763,42 @@ static directions dir_table[] = {
     {0xe0001, 0xe0001, direction::BN},  {0xe0020, 0xe007f, direction::BN},
     {0xe0100, 0xe01ef, direction::NSM}, {0xf0000, 0xffffd, direction::L},
     {0x100000, 0x10fffd, direction::L}};
+
+
 // CheckJoiners and CheckBidi are true for URL specification.
+
+
+inline static direction find_direction(uint32_t code_point) noexcept {
+    auto it = std::lower_bound(std::begin(dir_table), std::end(dir_table), code_point, 
+                  [](const directions& d, uint32_t c) { return d.final_code < c; });
+    
+    // next check is almost surely in vain, but we use it for safety.
+    if(it == std::end(dir_table)) { return direction::NONE; }
+    // We have that d.final_code >= c.
+    if (code_point >= it->start_code) { return it->direct; }
+    return direction::NONE;
+}
+
+inline static size_t find_last_not_of_nsm(const std::u32string_view label) noexcept {
+  for(size_t i = label.size() - 1;; i--) 
+    if(find_direction(label[i]) != direction::NSM) return i;
+
+  return std::u32string_view::npos;
+}
+
+// An RTL label is a label that contains at least one character of type R, AL, or AN.
+// https://www.rfc-editor.org/rfc/rfc5893#section-2
+inline static bool is_rtl_label(const std::u32string_view label) noexcept {
+  for (size_t i = 0; i < label.size(); i++) {
+    auto d = find_direction(label[i]);
+    if ((d == direction::R) || (d == direction::AL) || (d == direction::AN)) 
```

Comment:
moving this branch outside for loop, and using bit operations to manipulate a boolean value would make this function auto vectorizable.

Reply:
do you have any examples/articles/resources on this?

Reply:
I think I did it!